cpp/tests/utilities/cudf_test_utils.cuh

```diff
@@ -16,114 +16,69 @@
+               cudaMemcpyDeviceToHost);
 
-  for(size_t i = 0; i < num_rows; ++i)
-  {
-    // If the element is valid, print it's value
-    if(true == gdf_is_valid(h_mask.data(), i))
-    {
-      std::cout << h_data[i] << " ";
+    gdf_size_type const num_masks{gdf_get_num_chars_bitmask(num_rows)};
+    std::vector<gdf_valid_type> h_mask(num_masks, ~gdf_valid_type{0});
+    if (nullptr != the_column->valid) {
+      cudaMemcpy(h_mask.data(), the_column->valid,
+                 num_masks * sizeof(gdf_valid_type), cudaMemcpyDeviceToHost);
     }
-    // Otherwise, print an @ to represent a null value
-    else
-    {
-      std::cout << "@" << " ";
+
+    for (gdf_size_type i = 0; i < num_rows; ++i) {
+      // If the element is valid, print it's value
+      if (true == gdf_is_valid(h_mask.data(), i)) {
+        std::cout << h_data[i] << " ";
+      }
+      // Otherwise, print an @ to represent a null value
+      else {
+        std::cout << "@"
+                  << " ";
+      }
     }
+    std::cout << std::endl;
   }
-  std::cout << std::endl;
+};
 }
 
 void print_gdf_column(gdf_column const * the_column)
 {
-  const size_t num_rows = the_column->size;
-
-  const gdf_dtype gdf_col_type = the_column->dtype;
-  switch(gdf_col_type)
-  {
-    case GDF_INT8:
-      {
-        using col_type = int8_t;
-        col_type * col_data = static_cast<col_type*>(the_column->data);
-        print_typed_column<col_type>(col_data, the_column->valid, num_rows);
-        break;
-      }
-    case GDF_INT16:
-      {
-        using col_type = int16_t;
-        col_type * col_data = static_cast<col_type*>(the_column->data);
-        print_typed_column<col_type>(col_data, the_column->valid, num_rows);
-        break;
-      }
-    case GDF_INT32:
-      {
-        using col_type = int32_t;
-        col_type * col_data = static_cast<col_type*>(the_column->data);
-        print_typed_column<col_type>(col_data, the_column->valid, num_rows);
-        break;
-      }
-    case GDF_INT64:
-      {
-        using col_type = int64_t;
-        col_type * col_data = static_cast<col_type*>(the_column->data);
-        print_typed_column<col_type>(col_data, the_column->valid, num_rows);
-        break;
-      }
-    case GDF_FLOAT32:
-      {
-        using col_type = float;
-        col_type * col_data = static_cast<col_type*>(the_column->data);
-        print_typed_column<col_type>(col_data, the_column->valid, num_rows);
-        break;
-      }
-    case GDF_FLOAT64:
-      {
-        using col_type = double;
-        col_type * col_data = static_cast<col_type*>(the_column->data);
-        print_typed_column<col_type>(col_data, the_column->valid, num_rows);
-        break;
-      }
-    default:
-      {
-        std::cout << "Attempted to print unsupported type.\n";
-      }
-  }
+    cudf::type_dispatcher(the_column->dtype, column_printer{}, the_column);
 }
 
 /** ---------------------------------------------------------------------------*
```

Comment:
also inline `print_valid_data`. Or put in an anonymous namespace and call using another inline function. i can't add a comment there coz it's not part of the lines you touched.

Reply:
Ditto.